aten/src/ATen/native/Sorting.cpp

```diff
@@ -101,32 +90,99 @@ void quick_select_template(
       swap_fn(L + 1, L);
     }
 
-    i = L + 1;
-    j = R;
     piv = arr[L];
-    do {
-      do
-        i++;
-      while (gt_or_nan(piv, arr[i]));
-      do
-        j--;
-      while (gt_or_nan(arr[j], piv));
-      if (j < i)
-        break;
-      swap_fn(i, j);
-    } while (1);
-    swap_fn(L, j);
+    int64_t j = partition(L + 1, R, piv);
 
     // Re-set active partition
     if (j <= k)
-      L = i;
+      L = j;
     if (j >= k)
       R = j - 1;
   } while (1);
 }
 
 } // namespace
 
+// via SFINAE, this version will be used
+// if only a scalar partition function is available
+template <typename scalar_t, typename index_t>
+index_t partition(
+    scalar_t *values,
+    index_t *indices,
+    index_t L,
+    index_t R,
+    scalar_t piv,
+    bool largest,
+    long) {
+  auto cmp = largest ? [](scalar_t x, scalar_t y) {
+      return gt_or_nan<scalar_t>(x, y);
+    } : [](scalar_t x, scalar_t y) {
+      return gt_or_nan<scalar_t>(y, x);
+    };
+  auto swap = [&](index_t i, index_t j) {
+    std::swap(values[i], values[j]);
+    std::swap(indices[i], indices[j]);
+  };
+  return scalar_partition(values, cmp, swap, L, R, piv);
+}
+
+template <typename scalar_t, typename index_t>
+auto partition(
+    scalar_t *values,
+    index_t *indices,
+    index_t L,
+    index_t R,
+    scalar_t piv,
+    bool largest,
+    int) -> decltype(vec_qs_partition_inplace((scalar_t *)(nullptr), (scalar_t *)(nullptr), (index_t *)(nullptr), scalar_t(), bool()), index_t()){
```

Comment:
Ummmm... this is just an `int32_t`, right? Can we just say that?

Reply:
Oh, are these shenanigans because the type may change based on `index_t`? But I am still not sure why the return type cannot be straightforwardly written in terms of `index_t`...